builtin/logical/transit/path_decrypt.go

```diff
@@ -84,24 +109,115 @@ func (b *backend) pathDecryptWrite(
 		return logical.ErrorResponse("policy not found"), logical.ErrInvalidRequest
 	}
 
-	plaintext, err := p.Decrypt(context, nonce, ciphertext)
-	if err != nil {
-		switch err.(type) {
-		case errutil.UserError:
-			return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
-		case errutil.InternalError:
-			return nil, err
-		default:
-			return nil, err
+	batchInputRaw := d.Get("batch").(string)
+	var batchInputArray []BatchDecryptionItemRequest
+	var batchInput []byte
+	if len(batchInputRaw) != 0 {
+		batchInput, err = base64.StdEncoding.DecodeString(batchInputRaw)
+		if err != nil {
+			return logical.ErrorResponse("failed to base64-decode batch input"), logical.ErrInvalidRequest
+		}
+
+		if err := jsonutil.DecodeJSON([]byte(batchInput), &batchInputArray); err != nil {
+			return nil, fmt.Errorf("invalid input: %v", err)
+		}
+
+	} else {
+		ciphertext := d.Get("ciphertext").(string)
+		if len(ciphertext) == 0 {
+			return logical.ErrorResponse("missing ciphertext to decrypt"), logical.ErrInvalidRequest
+		}
+
+		batchInputArray = make([]BatchDecryptionItemRequest, 1)
+		batchInputArray[0] = BatchDecryptionItemRequest{
+			Ciphertext: ciphertext,
+			Context:    d.Get("context").(string),
+			Nonce:      d.Get("nonce").(string),
 		}
 	}
 
-	// Generate the response
-	resp := &logical.Response{
-		Data: map[string]interface{}{
-			"plaintext": plaintext,
-		},
+	if len(batchInputArray) == 0 {
+		return logical.ErrorResponse("missing input to process"), logical.ErrInvalidRequest
+	}
+
+	contextSet := true
```

Comment:
This code doesn't work either. It will fail if the last element in the array is missing context.

I still think you should set `contextSet` based on the first element. Just transform the two if loops into a single if/else:

```
if (item.Context == "" && contextSet) || (item.Context != "" && !contextSet) {
```

This should still probably be in a `switch` so you can skip it if the input array is length 1, since you have a zero case, a 1 case, and then a default case.